src/core/lib/event_engine/thread_pool/work_stealing_thread_pool.cc

```diff
@@ -0,0 +1,459 @@
+  // running instead of zero.
+  bool is_threadpool_thread = g_local_queue != nullptr;
+  thread_count()->BlockUntilThreadCount(CounterType::kLivingThreadCount,
+                                        is_threadpool_thread ? 1 : 0,
+                                        "shutting down", work_signal());
+  GPR_ASSERT(queue_.Empty());
+  quiesced_.store(true, std::memory_order_relaxed);
+  lifeguard_.BlockUntilShutdown();
+}
+
+bool WorkStealingThreadPool::WorkStealingThreadPoolImpl::SetThrottled(
+    bool throttled) {
+  return throttled_.exchange(throttled, std::memory_order_relaxed);
+}
+
+void WorkStealingThreadPool::WorkStealingThreadPoolImpl::SetShutdown(
+    bool is_shutdown) {
+  auto was_shutdown = shutdown_.exchange(is_shutdown);
+  GPR_ASSERT(is_shutdown != was_shutdown);
+  work_signal_.SignalAll();
+}
+
+void WorkStealingThreadPool::WorkStealingThreadPoolImpl::SetForking(
+    bool is_forking) {
+  auto was_forking = forking_.exchange(is_forking);
+  GPR_ASSERT(is_forking != was_forking);
+}
+
+bool WorkStealingThreadPool::WorkStealingThreadPoolImpl::IsForking() {
+  return forking_.load(std::memory_order_relaxed);
+}
+
+bool WorkStealingThreadPool::WorkStealingThreadPoolImpl::IsShutdown() {
+  return shutdown_.load(std::memory_order_relaxed);
+}
+
+bool WorkStealingThreadPool::WorkStealingThreadPoolImpl::IsQuiesced() {
+  return quiesced_.load(std::memory_order_relaxed);
+}
+
+void WorkStealingThreadPool::WorkStealingThreadPoolImpl::PrepareFork() {
+  SetForking(true);
+  thread_count()->BlockUntilThreadCount(CounterType::kLivingThreadCount, 0,
+                                        "forking", &work_signal_);
+  lifeguard_.BlockUntilShutdown();
+}
+
+void WorkStealingThreadPool::WorkStealingThreadPoolImpl::Postfork() {
+  SetForking(false);
+  Start();
+}
+
+// -------- WorkStealingThreadPool::WorkStealingThreadPoolImpl::Lifeguard
+// --------
+
+WorkStealingThreadPool::WorkStealingThreadPoolImpl::Lifeguard::Lifeguard()
+    : backoff_(grpc_core::BackOff::Options()
+                   .set_initial_backoff(kLifeguardMinSleepBetweenChecks)
+                   .set_max_backoff(kLifeguardMaxSleepBetweenChecks)
+                   .set_multiplier(1.3)) {}
+
+void WorkStealingThreadPool::WorkStealingThreadPoolImpl::Lifeguard::Start(
+    std::shared_ptr<WorkStealingThreadPoolImpl> pool) {
+  pool_ = std::move(pool);
+  grpc_core::Thread(
+      "lifeguard",
+      [](void* arg) {
+        auto* lifeguard = static_cast<Lifeguard*>(arg);
+        lifeguard->LifeguardMain();
+      },
+      this, nullptr,
+      grpc_core::Thread::Options().set_tracked(false).set_joinable(false))
+      .Start();
+}
+
+void WorkStealingThreadPool::WorkStealingThreadPoolImpl::Lifeguard::
+    LifeguardMain() {
+  thread_running_.store(true);
+  while (true) {
+    absl::SleepFor(absl::Milliseconds(
+        (backoff_.NextAttemptTime() - grpc_core::Timestamp::Now()).millis()));
```

Comment:
Why do we backoff exponentially here instead of having a fixed sleep duration ?

Reply:
Benchmarks and efficiency :-)

If the pool is highly active, we want a vigilant lifeguard because it will wake idle workers faster than the workers will wake themselves. If gRPC is idle, having the lifeguard wake up every 50 millis is needlessly expensive.

Reply:
Doesn't this mean that through line 267, the backoff timer keeps multiplying even for the case where there is some idle thread AND work to be done (i.e work queue is not empty) ? 

Shouldn't the backoff timer multiply only for the case where there are idle threads but no work needs to be done (i.e the pool is empty) ?

Reply:
Indeed, good catch. Fixed.